tracer/src/Datadog.Trace/Configuration/ImmutableTracerSettings.cs

```diff
@@ -508,13 +510,21 @@ internal string GetServiceName(Tracer tracer, string serviceName)
             {
                 return name;
             }
-            else if (MetadataSchemaVersion != SchemaVersion.V0)
+
+            if (MetadataSchemaVersion != SchemaVersion.V0)
             {
                 return tracer.DefaultServiceName;
             }
-            else
+
+            lock (_serviceNameCache)
```

Comment:
Just wondering, did you try this with a `ConcurrentDictionary`? Given this will be almost entirely read-only after the initial population, `ConcurrentDictionary` is optimised for lock-free reads, and that there could be contention for the lock in multi threaded scenario, I would have guessed concurrent dictionary would do better in throughput tests?

Reply:
I tried with ConcurrentDictionary and it was slower in all benchmarks I did. It's true benchmarks are single threaded but I didn't see any improvements in the throughput tests. That said, I would say you are theoretically right. :)

Reply:
Changed to `ConcurrentDictionary` in [84ce8de](https://github.com/DataDog/dd-trace-dotnet/pull/4125/commits/84ce8de1223471d73f08d97452e7fdf9cd483075) let's check the numbers again :)